src/lib/crypto/snarky_tests/snarky_tests.ml

```diff
@@ -0,0 +1,274 @@
+    let open Impl in
+    let%bind x = exists Boolean.typ ~compute:(As_prover.return true) in
+    let%bind y = exists Boolean.typ ~compute:(As_prover.return true) in
+    let%bind z = Boolean.(x &&& y) in
+    let%bind b3 = Boolean.(b1 && b2) in
+
+    let%bind () = Boolean.Assert.is_true z in
+    let%bind () = Boolean.Assert.is_true b3 in
+
+    Checked.return ()
+
+  let get_hash_of_circuit () =
+    let input_typ = Impl.Typ.tuple2 Impl.Boolean.typ Impl.Boolean.typ in
+    let return_typ = Impl.Typ.unit in
+    let cs : Impl.R1CS_constraint_system.t =
+      Impl.constraint_system ~input_typ ~return_typ main
+    in
+    let digest = Md5.to_hex (Impl.R1CS_constraint_system.digest cs) in
+    Format.printf "expected:\n%s\n\n" expected ;
+    Format.printf "obtained:\n%s\n" digest ;
+    assert (String.(digest = expected))
+end
+
+(* circuit-focused tests *)
+
+module BooleanCircuit = struct
+  module Request = struct
+    type _ Snarky_backendless.Request.t +=
+      | SimpleBool : bool Snarky_backendless.Request.t
+
+    let handler (Snarky_backendless.Request.With { request; respond }) =
+      match request with
+      | SimpleBool ->
+          respond (Provide true)
+      | _ ->
+          respond Unhandled
+  end
+
+  let input_typ = Impl.Typ.tuple2 Impl.Boolean.typ Impl.Boolean.typ
+
+  let return_typ = Impl.Typ.unit
+
+  let main ((b1, b2) : Impl.Boolean.var * Impl.Boolean.var) () =
+    let x =
+      Impl.exists Impl.Boolean.typ ~request:(fun _ -> Request.SimpleBool)
+    in
+    let y = Impl.exists Impl.Boolean.typ ~compute:(fun _ -> true) in
+
+    let z = Impl.Boolean.(x && y) in
+    let b3 = Impl.Boolean.(b1 && b2) in
+    Impl.Boolean.Assert.is_true z ;
+    Impl.Boolean.Assert.is_true b3 ;
+    ()
+end
+
+module FieldCircuit = struct
+  let input_typ = Impl.Field.typ
+
+  let return_typ = Impl.Typ.unit
+
+  let main (x : Impl.Field.t) () =
+    let y = Impl.Field.one in
+    let res = Impl.Field.(x + y) in
+    let two = Impl.Field.of_int 2 in
+
+    Impl.Field.Assert.equal res two ;
+    Impl.Field.Assert.not_equal res Impl.Field.one ;
+    Impl.Field.Assert.non_zero res
+end
+
+module RangeCircuits = struct
+  let input_typ = Impl.Field.typ
+
+  let return_typ = Impl.Typ.unit
+
+  let lte (x : Impl.Field.t) () =
+    Impl.Field.Assert.lte ~bit_length:2 x (Impl.Field.of_int 3) ;
+    ()
+
+  let gte (x : Impl.Field.t) () =
+    Impl.Field.Assert.gte ~bit_length:2 x (Impl.Field.of_int 3) ;
+    ()
+
+  let gt (x : Impl.Field.t) () =
+    Impl.Field.Assert.gt ~bit_length:2 x (Impl.Field.of_int 3) ;
+    ()
+
+  let lt (x : Impl.Field.t) () =
+    Impl.Field.Assert.lt ~bit_length:2 x (Impl.Field.of_int 3) ;
+    ()
+end
+
+module TernaryCircuit = struct
+  let input_typ = Impl.Boolean.typ
+
+  let return_typ = Impl.Typ.unit
+
+  let main (x : Impl.Boolean.var) () =
+    let two = Impl.Field.of_int 2 in
+    let res = Impl.if_ x ~typ:Impl.Field.typ ~then_:two ~else_:Impl.Field.one in
```

Comment:
Would be good to test a few different variations. For example, we should make sure that `~typ:(As_prover.Ref.typ ())` works